File: test/entities/btcblock_test.cpp
// Copyright (c) 2019-2020 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

#include "veriblock/entities/btcblock.hpp"

#include <gtest/gtest.h>

#include "veriblock/arith_uint256.hpp"
#include "veriblock/literals.hpp"

using namespace altintegration;

static const BtcBlock defaultBlock{
    536870912u,
    "f7de2995898800ab109af96779b979a60715da9bf2bbb745b300000000000000"_unhex,
    "f85486026bf4ead8a37a42925332ec8b553f8e310974fea1eba238f7cee6165e"_unhex,
    1555501858u,
    436279940u,
    (uint32_t)-1695416274};

static const std::string defaultBlockEncoded =
    "0000002000000000000000b345b7bbf29bda1507a679b97967f99a10ab0088899529def75e"
    "16e6cef738a2eba1fe7409318e3f558bec325392427aa3d8eaf46b028654f82213b75c841a"
    "011a2e00f29a";

TEST(BtcBlock, Deserialize) {
  auto blockEncoded = ParseHex(defaultBlockEncoded);
  auto stream = ReadStream(blockEncoded);
  auto decoded = BtcBlock::fromRaw(stream);

  EXPECT_EQ(decoded.version, defaultBlock.version);
  EXPECT_EQ(decoded.previousBlock.toHex(), defaultBlock.previousBlock.toHex());
  EXPECT_EQ(decoded.merkleRoot.toHex(), defaultBlock.merkleRoot.toHex());
  EXPECT_EQ(decoded.getBlockTime(), defaultBlock.getBlockTime());
  EXPECT_EQ(decoded.getDifficulty(), defaultBlock.getDifficulty());
  EXPECT_EQ(decoded.nonce, defaultBlock.nonce);

  EXPECT_FALSE(stream.hasMore(1)) << "stream has more data";
}

TEST(BtcBlock, Serialize) {
  WriteStream stream;
  defaultBlock.toRaw(stream);
  auto btcBytes = stream.data();
  auto blockEncoded = HexStr(btcBytes);
  EXPECT_EQ(blockEncoded, defaultBlockEncoded);
}

TEST(BtcBlock, RoundTrip) {
  auto blockEncoded = ParseHex(defaultBlockEncoded);
  auto stream = ReadStream(blockEncoded);
  auto decoded = BtcBlock::fromRaw(stream);
  EXPECT_EQ(decoded.version, defaultBlock.version);

  WriteStream outputStream;
  decoded.toRaw(outputStream);
  auto btcBytes = outputStream.data();
  auto blockReEncoded = HexStr(btcBytes);
  EXPECT_EQ(blockReEncoded, defaultBlockEncoded);
}

TEST(BtcBlock, RoundTripNew) {
  auto blockEncoded = ParseHex(defaultBlockEncoded);
  BtcBlock decoded;
  ValidationState state;
  bool ret = DeserializeRaw(blockEncoded, decoded, state);
  ASSERT_TRUE(ret);
  EXPECT_TRUE(state.IsValid());
  EXPECT_EQ(decoded.version, defaultBlock.version);

  WriteStream outputStream;
  decoded.toRaw(outputStream);
  auto btcBytes = outputStream.data();
  auto blockReEncoded = HexStr(btcBytes);
  EXPECT_EQ(blockReEncoded, defaultBlockEncoded);
}

TEST(BtcBlock, getBlockHash_test) {
  BtcBlock block;
  block.version = 536870912;
  block.previousBlock = uint256(
      "00000000000000b345b7bbf29bda1507a679b97967f99a10ab0088899529def7"_unhex);
  block.merkleRoot = uint256(
      "5e16e6cef738a2eba1fe7409318e3f558bec325392427aa3d8eaf46b028654f8"_unhex);
  block.timestamp = 1555501858;
  block.bits = 436279940;
  block.nonce = 2599551022;

  EXPECT_EQ(
      ArithUint256::fromLEBytes(block.getHash()),
      ArithUint256::fromHex(
          "000000000000000246200f09b513e517a3bd8c591a3b692d9852ddf1ee0f8b3a"));
}

File: bindings/go/entities/btcblock_test.go
// Copyright (c) 2019-2021 Xenios SEZC
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package entities

import (
	"bytes"
	"encoding/hex"
	"testing"

	"github.com/stretchr/testify/assert"
)

var (
	defaultBtcBlock = BtcBlock{
		536870912,
		parse32Bytes("f7de2995898800ab109af96779b979a60715da9bf2bbb745b300000000000000"),
		parse32Bytes("f85486026bf4ead8a37a42925332ec8b553f8e310974fea1eba238f7cee6165e"),
		1555501858,
		436279940,
		2599551022,
	}
	defaultBtcBlockEncoded = "0000002000000000000000b345b7bbf29bda1507a679b97967f99a10ab0088899529def75e16e6cef738a2eba1fe7409318e3f558bec325392427aa3d8eaf46b028654f82213b75c841a011a2e00f29a"
)

func parseHex(src string) []byte {
	res, _ := hex.DecodeString(src)
	return res
}

func parse32Bytes(src string) [32]byte {
	buf := parseHex(src)
	var block [32]byte
	copy(block[:], buf)
	return block
}

func TestBtcBlockDeserialize(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	blockEncoded := parseHex(defaultBtcBlockEncoded)
	stream := bytes.NewReader(blockEncoded)
	decoded := BtcBlock{}
	assert.NoError(decoded.FromRaw(stream))

	assert.Equal(defaultBtcBlock.Version, decoded.Version)
	assert.Equal(defaultBtcBlock.PreviousBlock, decoded.PreviousBlock)
	assert.Equal(defaultBtcBlock.MerkleRoot, decoded.MerkleRoot)
	assert.Equal(defaultBtcBlock.Timestamp, decoded.Timestamp)
	assert.Equal(defaultBtcBlock.Bits, decoded.Bits)
	assert.Equal(defaultBtcBlock.Nonce, decoded.Nonce)

	assert.Equal(0, stream.Len(), "Stream has more data")
}

func TestBtcBlockSerialize(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	stream := new(bytes.Buffer)
	assert.NoError(defaultBtcBlock.ToRaw(stream))
	blockEncoded := hex.EncodeToString(stream.Bytes())
	assert.Equal(defaultBtcBlockEncoded, blockEncoded)
}

func TestBtcBlockRoundTrip(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	blockEncoded := parseHex(defaultBtcBlockEncoded)
	stream := bytes.NewReader(blockEncoded)
	decoded := BtcBlock{}
	assert.NoError(decoded.FromRaw(stream))
	assert.Equal(defaultBtcBlock.Version, decoded.Version)

	outputStream := new(bytes.Buffer)
	assert.NoError(decoded.ToRaw(outputStream))
	blockReEncoded := hex.EncodeToString(outputStream.Bytes())
	assert.Equal(defaultBtcBlockEncoded, blockReEncoded)
}

func TestBtcBlockGetBlockHash(t *testing.T) {
	t.Parallel()

	assert := assert.New(t)

	blockEncoded := parseHex(defaultBtcBlockEncoded)
	stream := bytes.NewReader(blockEncoded)
	block := BtcBlock{}
	assert.NoError(block.FromRaw(stream))

	hash := block.GetHash()
	assert.Equal("ebaa22c5ffd827e96c4450ad5dd35dbec2aa45e15cdb5ce9928f543f4cebf10e", hex.EncodeToString(hash[:]))
}
